### ISSUE TITLE
[Question] Deny TURN own peer IP

### ISSUE THREAD

**Issue by cyberaa** [coturn/coturn#1407]
Hey guys,

I have the following configuration:

```
listening-port=80
tls-listening-port=443
no-tlsv1
no-tlsv1_1
listening-ip=10.0.10.5
external-ip=34.100.10.47
fingerprint
lt-cred-mech
user=myuser:mypass
realm=turn.test.com
cert=/etc/certs/fullchain.pem
pkey=/etc/certs/privkey.pem
cipher-list="TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384..."
dh-file=/etc/ssl/private/dhparams.pem
no-multicast-peers
no-loopback-peers
no-software-attribute
...
denied-peer-ip=0.0.0.0-0.255.255.255
denied-peer-ip=10.0.0.0-10.255.255.255 # Denies own IP 10.0.10.5
```

**Questions:**

* Is it fine to deny TURN's own IP(10.0.10.5)?
* Will this still work? 
* Is there any scenario, in which, this will fail?

**Reasoning:**

I have done this to prevent SSRF attacks by using **stunner** to tunnel unauthorized HTTP requests to the TURN server.

**Observations:**

* I have tested with Tricke ICE tool and I have positive results, however, using other tools like test.8x8.vc, it reported failure in the **relay connectivity** but not **host connectivity**.
* I also have seen other configurations that deny TURN own IP.

**Comment by eakraly** [comment]
Hi @cyberaa !
There are a few good resources on what you are wondering about. 
- https://www.rtcsec.com/article/slack-webrtc-turn-compromise-and-bug-bounty/
- https://www.rtcsec.com/article/cve-2020-26262-bypass-of-coturns-access-control-protection/

The reason to block is to prevent an attack that tries to connect to a port on the machine running coturn (some other service, internal service, etc) as you correctly figured out.

On the other hand, double TURN on the same server requires you to be able to connect to self.

The correct version is to configure coturn in a way that prevents connection to non-public addresses. It seems like that is exactly what you did.

As for the failure in test.8x8.vc - I suggest to disable all the "limitations" and re-add them one by one until you find what breaks it

**Comment by DenysGonchar** [comment]
TL;DR:
 
* `Is it fine to deny TURN's own IP(10.0.10.5)?` -  no, it's not fine
* `Will this still work?` - it won't in some cases, like with [test.8x8.vc](https://test.8x8.vc) 
* `Is there any scenario, in which, this will fail?` - It fails when two clients require relaying through the same Coturn server.

I’m not an expert on STUN/TURN, so it would be great if someone from the Coturn development team could confirm this, but after a quick look at the code I believe I understand what’s happening:

- This appears to be a typical scenario where two clients use the same STUN/TURN server because they are provisioned by the same application server.
- When both clients require relaying, their traffic is routed internally within the server.
- The case of [test.8x8.vc](https://test.8x8.vc) fits this pattern, as it uses the same configuration for both clients.
- This approach is efficient and helps reduce network bandwidth usage. However, when `--external-ip` is configured, both clients attempt to route traffic via that external IP.
- Coturn handles this by creating an internal mapping from the external IP to the corresponding internal relay IPs during initialization. As a result, traffic is not actually sent to the external IP but redirected to the mapped internal address.
- Even though this mapped address is internal and traffic stays within the server, it is still validated against the allowed and denied peer lists.
- Therefore, you need to explicitly whitelist the internal relay IP address using the --allowed-peer-ip option.
- Alternatively, you can specify external IP in the external_ip/internal_ip format (e.g., `--external-ip=1.2.3.4/10.0.0.1`), and Coturn will handle the whitelisting automatically.
- Please note that whitelisting the internal relay IP address may expose some other services that run on the same machine as Coturn.

It might be worth documenting this in the `turnserver`'s man page.

Here are the references in the code:

- Mapping with whitelisting: https://github.com/coturn/coturn/blob/4.9.0/src/apps/relay/mainrelay.c#L2185-L2188
- Mapping without whitelisting: https://github.com/coturn/coturn/blob/4.9.0/src/apps/relay/mainrelay.c#L3365

BTW you can run a similar test with `turnutils_uclient` tool:
```
turnutils_uclient -y -n 10 -u <user_name>  -w <password> <TURN_server_IP>
```
